tianshou/data/batch.py

```diff
@@ -174,20 +198,213 @@ def _alloc_by_keys_diff(
             meta[key] = _create_value(batch[key], size, stack)
 
 
-class Batch:
+# Note: This is implemented as a protocol because the interface
+# of Batch is always extended by adding new fields. Having a hierarchy of
+# protocols building off this one allows for type safety and IDE support despite
+# the dynamic nature of Batch
+@runtime_checkable
```

Comment:
It allows `isnistance` to be used on protocols, really handy! I can also make the other protocols checkable.

```python
from typing import Protocol, runtime_checkable

class AProtocol:
    a: str
    
@runtime_checkable
class CheckableAProtocol(Protocol):
    a: str
    
class A:
    a = "a"
    
class B:
    b = "b"
    
isinstance(A(), AProtocol)
Out[7]: False
isinstance(A(), CheckableAProtocol)
Out[8]: True
isinstance(B(), CheckableAProtocol)
Out[9]: False
```

Reply:
> make the other protocols checkable.

good idea! But will it introduce extra runtime overhead?

Reply:
Indeed, isinstance on runtime-checked protocols is slower. Since for now it's not needed almost anywhere, certainly not in critical loops, it's not a problem. But maybe introducing @runtime_checkable everywhere without an explicit need for it sends the wrong message